test/functional/shared.js

```diff
@@ -276,6 +276,116 @@ class APMEventCollector {
   }
 }
 
+/**
+ * makes documents can take an array of documents or a count / shape and create
+ * then progamatically create documents
+ *
+ * @example
+ * docMaker.bind(this)({ count: 10, shape: (x) => ({x}) })
+ *
+ * @example
+ * docMaker.bind(this)([{a:1}])
+ */
+function docMaker(collection, docs, callback) {
+  if (Array.isArray(docs)) {
+    return collection.insertMany(docs, this.configuration.writeConcernMax(), callback);
+  } else if (docs.count && docs.shape) {
+    var total = docs.count;
+    for (var i = 0; i < docs.count; i++) {
+      collection.insertOne(docs.shape(i), this.configuration.writeConcernMax(), e => {
+        expect(e).to.not.exist;
+        total = total - 1;
+        if (total === 0) callback();
+      });
+    }
+  } else {
+    return callback();
+  }
+}
+
+// TODO: [THOMAS] Will remove just wanted to ensure that we dont want to test cursor on topology / server directly
+// function withCursorFromServer(options, callback) {
+//   const defaultOptions = {
+//     docs: [],
+//     database: 'integration_tests',
+//     collection: 'cursor',
+//     readPreference: 'primary',
+//     writeConcern: { w: 1 },
+//     ordered: true,
+//     batchSize: undefined
+//   };
+//   const opts = { ...defaultOptions, ...options };
+//   if (opts.collection === defaultOptions.collection) {
+//     throw new Error('please set a collection');
+//   }
+//   opts.ns = `${opts.collection}.${opts.database}`;
+//   const topology = this.configuration.newTopology();
+//   return topology.connect(err => {
+//     expect(err).to.not.exist;
+//     this.defer(() => topology.close());
+//     return topology.selectServer(opts.readPreference, (err, server) => {
+//       this.defer(() => server.destroy());
+//       expect(err).to.not.exist;
+//       const { ns, docs, writeConcern, ordered, batchSize } = opts;
+//       return server.insert(ns, docs, { writeConcern, ordered }, (err, results) => {
+//         expect(err).to.not.exist;
+//         expect(results.n).to.equal(docs.length);
+//         const cursor = topology.cursor(ns, {
+//           find: ns.split('.')[1],
+//           filter: {},
+//           batchSize: batchSize || docs.length
+//         });
+//         return callback(cursor);
+//       });
+//     });
+//   });
+// }
+
+function withCursorFromClient(optionsOrCallback, callback) {
```

Comment:
I would really like to avoid writing so many specialized `withXXX` methods. `withClient` makes sense because there's a decent amount of configuration, but otherwise I would recommend just using `this.defer` or just explicitly closing cursors when you're done with them. `withCursorFromClient` currently connects a client, optionally monitors commands, creates a collection, inserts documents into it, and then calls a lambda with a bunch of things.. very magical

We must not be afraid of code duplication in our tests! It is meant to be code written like a user so it actually improves readability and maintainability to have clear steps, even if they are copy-pasta.

Reply:
Rather than use the `withXXX` I create a vscode snippet. May be something we do for specific repetitive tests? Could be a good middleground?

```ts
const BATCH_SIZE = 2;
const COLLECTION = 'cursor0';
const DOCS = [{ a: 1 }, { a: 2 }, { a: 3 }];

const testCursor = cursor => {
  $1
};

const configuration = this.configuration;
const client = configuration.newClient(configuration.writeConcernMax(), { poolSize: 1 });
return client.connect((err, client) => {
  expect(err).to.not.exist;
  this.defer(() => client.close());
  const db = client.db(configuration.db);
  return db.createCollection(COLLECTION, (err, collection) => {
    expect(err).to.not.exist;
    return collection.insertMany(DOCS, this.configuration.writeConcernMax(), err => {
      expect(err).to.not.exist;
      const cursor = collection.find({});
      this.defer(() => cursor.close());
      cursor.batchSize(BATCH_SIZE);
      return testCursor(cursor);
    });
  });
});
```